### DISCUSSION TITLE
Question: Keeping UIView instances outside the closure

### DISCUSSION THREAD

**Discussion by oriolpregona** [siteline/SwiftUI-Introspect#231] [Q&A] [answered]
This is more a technical question than an issue, my approach worked really well so far, but I wanted to know if this would overstretch the purpose of this library, and see if there could be derived issues that I didn't foresee.

I need to implement fairly complicated behaviours on a ScrollView and so far none of these are available in pure SwiftUI, so I kept an instance of the scrollView as in the example below, assigning its delegate and using KVO on it, so I can keep track/change the contentOffset, detect when the user did release the dragging gesture and more.
```
        ScrollView {
            MyListView()
        }
        .introspectScrollView { scrollView in   
            self.myScrollView = scrollView
        }
```
Note this code is a simplified version, I noticed that the actual scrollView instance provided sometimes changes and I keep always the latest version, and I maintain the delegate, KVO and all properties updated accordingly.
It feels a bit hacky, but so far it works great, no crashes, good performance, no unexpected failures.

Is this approach reliable enough?

Thanks!

**Comment by davdroman** [comment, marked as answer]
@oriolpregona this is totally fine. I use an approach similar to yours for `UINavigationController` in my https://github.com/davdroman/swiftui-navigation-transitions library.

As a side note, note there's a [newly released](https://github.com/siteline/SwiftUI-Introspect/releases/tag/0.4.0) module that you should if possible prefer using, as follows:

```swift
import SwiftUIIntrospect

ScrollView {
    MyListView()
}
.introspect(.scrollView, on: .iOS(.v13, .v14, .v15, .v16), .tvOS(.v13, .v14, .v15, .v16)) { scrollView in
    self.myScrollView = scrollView
}
```

**Reply by davdroman** [reply]
Small update on this: `@Weak` now ships with SwiftUIIntrospect out of the box in the latest 0.12.0 release. [More info](https://github.com/siteline/SwiftUI-Introspect?tab=readme-ov-file#keep-instances-outside-the-customize-closure).

**Comment by hiETsang** [comment]
I happen to need such a feature as well, but I don't know how to implement it. Could you please teach me how to write the code for this solution? Thank you very much.

**Reply by davdroman** [reply]
Here's the safest way to do it:

```swift
import SwiftUI
import SwiftUIIntrospect

@propertyWrapper
final class Weak<T: AnyObject> {
    private weak var _wrappedValue: T?

    var wrappedValue: T? {
        get { _wrappedValue }
        set { _wrappedValue = newValue }
    }

    init(wrappedValue: T? = nil) {
        self._wrappedValue = wrappedValue
    }
}

struct ContentView: View {
    @Weak var scrollView: UIScrollView?

    var body: some View {
        ScrollView {
            // ...
        }
        .introspect(.scrollView, on: .iOS(.v13, .v14, .v15, .v16, .v17)) { scrollView in
            self.scrollView = scrollView
        }
    }
}
```

Using `@Weak` instead of SwiftUI's `@State` prevents nasty memory leaks.

I'll update the README before 1.0 to include this and other advanced use cases.

**Reply by davdroman** [reply]
Small update on this: `@Weak` now ships with SwiftUIIntrospect out of the box in the latest 0.12.0 release. [More info](https://github.com/siteline/SwiftUI-Introspect?tab=readme-ov-file#keep-instances-outside-the-customize-closure).
